op.c

```diff
@@ -2069,16 +2070,69 @@ Perl_scalar(pTHX_ OP *o)
                         )
                 )
                 {
+
+                    if (OP_TYPE_IS(o, OP_LIST) && !op_parent(o)) {
+                        /* Is the list now just an obvious scalar pushop?
+                         *     <@> list sKP ->6
+                         *         <0> pushmark v ->4
+                         *         <$> const(IV 3) s ->5
+                         */
+                        OP* first = cLISTOPo->op_first;
+                        assert(OP_TYPE_IS(first, OP_PUSHMARK));
+                        OP* sib1 = OpSIBLING(first);
+                        assert(sib1);
+                        OP* sib2 = OpSIBLING(sib1);
+                        if (!sib2) {
+                            if (
+                                PL_opargs[sib1->op_type] & OA_RETSCALAR
+                            ){
+                                assert(sib1->op_next == sib1);
+                                /* Yup. The PUSHMARK and LIST are redundant.
+                                 * They can be stripped out. */
+                                op_sibling_splice(o,first,1,NULL);
+                                op_free(o);
+                                return sib1;
+                            }
+                        }
+                    }
+
                     /* tail call optimise calling scalar() on the last kid */
+                    assert(kid);
                     next_kid = kid;
                     goto do_next;
                 }
                 else if (kid->op_type == OP_LEAVEWHEN)
                     scalar(kid);
-                else
+                else {
                     scalarvoid(kid);
+
+                    if (OP_TYPE_IS(kid, OP_NULL) && !(kid->op_flags & OPf_KIDS)
+                        && prev_kid
+                      ) {
+                        /* This OP is now defunct. Strip it out. */
+                        if (kid->op_next == kid || kid->op_next == sib) {
+                            if (prev_kid->op_next == kid)
+                                prev_kid->op_next = kid->op_next;
+
+                            prev_kid->op_sibparent = kid->op_sibparent;
+                            op_free(kid); kid = NULL;
+
+                            /* A NEXTSTATE with no sibling OPs is redundant
+                             * if another NEXTSTATE follows it. Null it out
+                             * rather than removing it, in case anything needs
+                             * to probe it for file/line/hints info. */
+                            if (OP_TYPE_IS(prev_kid, OP_NEXTSTATE) && sib
+                                && OP_TYPE_IS(sib, OP_NEXTSTATE)) {
+                                op_null(prev_kid);
```

Comment:
Don't forget that under `perl -d`, these are `OP_DBSTATE`. I wrote a handy helper macro for such; `OP_TYPE_IS_COP_NN`.

```c
if (OP_TYPE_IS_COP_NN(prev_kid) && OP_TYPE_IS_COP_NN(sib))
```